Not found page and routing logic

cikzh opened the pull request:
This PR adds a Not Found page and the logic to trigger the page.

It makes use of the `errorElement` logic in `react-router-dom`. When an unhandled error is thrown, the component defined at `errorElement` will be rendered instead. This way, we can easily define different kinds of error pages in different parts of the application.

Also adds an e2e test to check if the Not Found page is rendered when visiting a page with a non-existing election ID, it also checks if the browser URL remains the same. I'm not sure if it is possible to do this in vitest. If it is, please let me know how and I'll move it.

jschuurk-kr commented:
> Also adds an e2e test to check if the Not Found page is rendered when visiting a page with a non-existing election ID, it also checks if the browser URL remains the same. I'm not sure if it is possible to do this in vitest.

Should be possible with React's `MemoryRouter` and `router.state.location.pathname`: https://stackoverflow.com/a/73730116

cikzh commented:
Awesome, thanks! I'll look into it

cikzh commented:
Thanks @jschuurk-kr ! The test is now done in vitest

praseodym commented on frontend/app/routes.test.tsx:
```diff
@@ -0,0 +1,44 @@
+import { createMemoryRouter, RouterProvider } from "react-router-dom";
+
+import { render } from "@testing-library/react";
+import { describe, expect, test, vi } from "vitest";
+
+import { screen } from "app/test/unit";
+
+import { ApiProvider } from "@kiesraad/api";
+
+import { routes } from "./routes";
+
+const setupTestRouter = () => {
+  return createMemoryRouter(routes);
+};
+
+describe("routes", () => {
+  test("Non existing election id results in not found page", async () => {
+    // Since we test what happens after an error, we want vitest to ignore them
+    vi.spyOn(console, "error").mockImplementation(() => {
+      /* do nothing */
+    });
+
+    const router = setupTestRouter();
+
+    // Router sanity checks
+    expect(router.state.location.pathname).toEqual("/");
+    await router.navigate("/1/input");
+    expect(router.state.location.pathname).toEqual("/1/input");
+
+    // Navigate to a non-existing page
+    await router.navigate("/-1/input");
+
+    // NOTE: We're not using the wrapped render function here,
+    // since we want control over our own memory router.
+    render(
```
Maybe clarify that this not the `render` function from `app/test/unit` by renaming the import?

```suggestion
import { render as rtlRender } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";

import { screen } from "app/test/unit";

import { ApiProvider } from "@kiesraad/api";

import { routes } from "./routes";

const setupTestRouter = () => {
  return createMemoryRouter(routes);
};

describe("routes", () => {
  test("Non existing election id results in not found page", async () => {
    // Since we test what happens after an error, we want vitest to ignore them
    vi.spyOn(console, "error").mockImplementation(() => {
      /* do nothing */
    });

    const router = setupTestRouter();

    // Router sanity checks
    expect(router.state.location.pathname).toEqual("/");
    await router.navigate("/1/input");
    expect(router.state.location.pathname).toEqual("/1/input");

    // Navigate to a non-existing page
    await router.navigate("/-1/input");

    // NOTE: We're not using the wrapped render function here,
    // since we want control over our own memory router.
    rtlRender(
```

praseodym commented:
I noticed that frontend with MSW doesn't show 404 pages because the mock API doesn't check the election id. I think it should be easy enough to return a 404 from the mock API for invalid election ids, could you add this?

cikzh commented:
Yes, I noticed too, I got around it by using `overrideOnce`, but changing the MSW makes indeed more sense. Will do.

On a side note: I'm fixing the integer parsing issues in the `Layout` elements. I wish there was a way to do this in the router, as this problem will be present in other `Layout` objects as wel, but as far as I understand from the docs, they removed the ability to do so in a recent version. Is there a way to check if a path segment is numeric in `react-router-dom` that we know?

cikzh commented:
Yes, I found that too and I'm implementing it similar to that now.

cikzh commented:
>     6. `http://0.0.0.0:8080/1/input/1/list/` shows the 404 page, meaning we have some dead code in `frontend/app/module/input/PollingStation/CandidatesVotesPage.tsx` trying to display "Geen lijstnummer gevonden" for this case.
> 
> 
> I think 1, 3, 6, need to be addressed. 4 is as expected right now because #108.
> 
> No 3. is caused by `parseInt()`: _"If parseInt encounters a character that is not a numeral in the specified radix, it ignores it and all succeeding characters and returns the integer value parsed up to that point."_

@jschuurk-kr I added test cases for 1 and 3 and fixed them. I think 6 showing the 404 page is correct, as it is an invalid route. The route should always contain at least a list number.

jschuurk-kr commented:
> I think 6 showing the 404 page is correct, as it is an invalid route. The route should always contain at least a list number.

Agree that showing the 404 page is correct. Was just wondering if that means that the code in `frontend/app/module/input/PollingStation/CandidatesVotesPage.tsx` that displays "Geen lijstnummer gevonden" can no longer be reached and if so, should be removed.

cikzh commented:
Ah yes, that one is indeed dead as it's handled by the router. I'll remove it :+1:

jschuurk-kr commented on frontend/app/module/NotFoundProvider.tsx:
```diff
@@ -0,0 +1,29 @@
+  },
+});
+
+export const NotFoundProvider = (props: object) => {
```
My IDE can't find any usages of the `NotFoundProvider`.

cikzh replied:
Funny, I would expect the linter to catch this. Since we use the `errorElement` logic of the router, we don't need a provider or context at all. Thanks!

jschuurk-kr replied:
Looks like we need [`eslint-plugin-import`](https://www.npmjs.com/package/eslint-plugin-import) for that, because `eslint` only does unused vars and unused function params.

praseodym commented:
Using a non-existent polling station in the MSW version doesn't lead to the 404 page (scenario 4 in https://github.com/kiesraad/abacus/pull/193#issuecomment-2255553886), is this in scope?

cikzh commented:
It looks to be a bug in getting the polling station. I'll fix it in this PR

cikzh commented:
[4911e73](https://github.com/kiesraad/abacus/pull/193/commits/4911e730b92e61e2225a80759f33b3bde1d82480) removes loading messages as discussed here: https://github.com/kiesraad/abacus/issues/108#issuecomment-2257565164

Lionqueen94 commented on frontend/app/module/input/PollingStation/PollingStationLayout.tsx:
```diff
@@ -1,82 +1,75 @@
-        <main>
+    <PollingStationFormController
+      election={election}
+      pollingStationId={parseInt(pollingStationId || "0")}
```
```suggestion
      pollingStationId={pollingStation.id}
```

cikzh replied:
Yes, good one, we parse this unnecessarily

Lionqueen94 commented on frontend/lib/util/hook/useNumericParam.ts:
```diff
@@ -0,0 +1,18 @@
+import { useParams } from "react-router-dom";
+
+//
```
Missing comment

cikzh replied:
No comment.

rnijveld commented on frontend/lib/util/hook/useNumericParam.ts:
```diff
@@ -0,0 +1,23 @@
+
+  // Matches only if the whole string is numeric
+  const parsedParam = parseIntStrict(param);
+  if (!parsedParam) {
```
Technically this doesn't work when the integer is `0` right now. Shouldn't be a problem for all the id params we have right now, given they start at `1`, but for other numeric values it might be an issue.

praseodym replied:
Maybe this then?
```suggestion
  if (parsedParam !== undefined) {
```

cikzh replied:
[8e29b14](https://github.com/kiesraad/abacus/pull/193/commits/8e29b148112e110059702cb33b1cd246bc8dfedf) addresses this
